optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,344 @@
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = trans.bounds[:, 0] + sample * (trans.bounds[:, 1] - trans.bounds[:, 0])
+        return trans.untransform(sample[0, :])
+
+    def after_trial(
+        self,
+        study: "optuna.Study",
+        trial: "optuna.trial.FrozenTrial",
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        self._independent_sampler.after_trial(study, trial, state, values)
+
+    def _sample_qmc(
+        self, study: Study, search_space: Dict[str, BaseDistribution]
+    ) -> numpy.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, search_space)
+        d = len(search_space)
+
+        # Use cached `qmc_engine` or construct a new one.
+        if self._is_engine_cached(d, sample_id):
+            qmc_engine: scipy.stats.qmc.QMCEngine = self._cached_qmc_engine
+        else:
+            if self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "halton" and "sobol".'
+                )
+                raise ValueError(message)
+
+        forward_size = sample_id - qmc_engine.num_generated  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
+        sample = qmc_engine.random(1)
+        self._cached_qmc_engine = qmc_engine
+
+        return sample
+
+    def _find_sample_id(self, study: Study, search_space: Dict[str, BaseDistribution]) -> int:
+
+        qmc_id = ""
+        qmc_id += self._qmc_type
+        qmc_id += str(search_space)
+        # Sobol/Halton sequences without scrambling do not use seed.
+        if self._scramble:
+            qmc_id += str(self._seed)
+        hashed_qmc_id = hash(qmc_id)
+        key_qmc_id = f"qmc ({hashed_qmc_id})'s last sample id"
+
+        # TODO(kstoneriv3): Following try-except block assumes that the block is
+        # an atomic transaction. Without this assumption, current implementation
+        # only ensures that each `sample_id` is sampled at least once.
```

Comment:
Actually, it is not an atomic transaction. Is it okay?

Reply:
Atomicity is unnecessary for single-thread optimization but in parallel / distributed cases, it makes a difference. As long as the evaluation of function takes a much longer time than the access to the storage here, the transaction becomes "almost" atomic. Experimentally, these operations are almost atomic unless the function is super easy to evaluate and the number of workers is more than 100. If this is not atomic, multiple workers will end up evaluating the same `sample_id` (therefore, same hyperparameters), but this does not result in skip of `sample_id` anyways. Therefore, we don't actually need atomicity, but without atomicity, the performance of optimization might decay due to duplicated evaluation of the hyperparameters.